File: tools/publisher/public_listing_dedupe.py
"""Conservative public-listing dedupe for adjacent identical source emissions."""
from __future__ import annotations

from datetime import datetime, timezone
import re
import unicodedata


DUPLICATE_WINDOW_SECONDS = 10 * 60
_ZERO_WIDTH = re.compile(r"[\u200b-\u200f\u202a-\u202e\u2060\ufeff]")


def _timestamp(value: object) -> float:
    raw = str(value or "").strip().replace(" ", "T")
    if not raw:
        return 0.0
    try:
        return datetime.fromisoformat(raw).replace(tzinfo=timezone.utc).timestamp()
    except ValueError:
        return 0.0


def _normalized_text(value: object) -> str:
    return re.sub(r"\s+", " ", _ZERO_WIDTH.sub("", unicodedata.normalize("NFKC", str(value or "")))).strip()


def public_listing_key(post: dict) -> str:
    media_identity = str(
        post.get("media_unique_id")
        or post.get("photo_unique_id")
        or post.get("media_file_id")
        or post.get("photo_file_id")
        or ""
    ).strip()
    return _normalized_text(post.get("text")) + "\0" + media_identity


def dedupe_public_listing_posts(posts: list[dict]) -> list[dict]:
    ordered = sorted(posts, key=lambda post: (_timestamp(post.get("created_at")), int(post.get("id") or 0)), reverse=True)
    last_seen: dict[str, float] = {}
    result: list[dict] = []
    for post in ordered:
        key = public_listing_key(post)
        current = _timestamp(post.get("created_at"))
        previous = last_seen.get(key, 0.0)
        adjacent_duplicate = current > 0 and previous > 0 and abs(previous - current) <= DUPLICATE_WINDOW_SECONDS
        if not adjacent_duplicate:
            result.append(post)
        if current > 0:
            last_seen[key] = current
    return result

File: website/dreary-disk/src/lib/publicListingDedupe.ts
export type PublicListingPost = {
  id: number;
  slug?: string | null;
  text?: string | null;
  created_at?: string | null;
  media_unique_id?: string | null;
  photo_unique_id?: string | null;
  media_file_id?: string | null;
  photo_file_id?: string | null;
};

const PUBLIC_LISTING_DUPLICATE_WINDOW_MS = 10 * 60 * 1000;

function timestamp(value: unknown): number {
  const parsed = Date.parse(String(value || "").replace(" ", "T"));
  return Number.isFinite(parsed) ? parsed : 0;
}

function normalizedListingText(value: unknown): string {
  return String(value || "")
    .normalize("NFKC")
    .replace(/[\u200B-\u200F\u202A-\u202E\u2060\uFEFF]/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

function publicListingKey(post: PublicListingPost): string {
  const mediaIdentity = String(
    post.media_unique_id || post.photo_unique_id || post.media_file_id || post.photo_file_id || ""
  ).trim();
  return normalizedListingText(post.text) + "\u0000" + mediaIdentity;
}

export function sortPublicListingPosts<T extends PublicListingPost>(posts: T[]): T[] {
  return [...posts].sort((left, right) =>
    timestamp(right.created_at) - timestamp(left.created_at) || right.id - left.id
  );
}

export function dedupePublicListingPosts<T extends PublicListingPost>(posts: T[]): T[] {
  const sorted = sortPublicListingPosts(posts);
  const lastSeen = new Map<string, number>();
  const result: T[] = [];

  for (const post of sorted) {
    const key = publicListingKey(post);
    const current = timestamp(post.created_at);
    const previous = lastSeen.get(key) || 0;
    const isAdjacentSourceDuplicate =
      current > 0 && previous > 0 && Math.abs(previous - current) <= PUBLIC_LISTING_DUPLICATE_WINDOW_MS;

    if (!isAdjacentSourceDuplicate) result.push(post);
    if (current > 0) lastSeen.set(key, current);
  }

  return result;
}
